src/conn.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use rusqlite;
+
+use edn;
+use errors::*;
+use mentat_core::{
+    Schema,
+    TypedValue,
+};
+use mentat_db::db;
+use mentat_db::{
+    transact,
+    PartitionMap,
+    TxReport,
+};
+use mentat_tx_parser;
+use query::{
+    q_once,
+    QueryResults,
+};
+
+
+/// Connection metadata required to query from, or apply transactions to, a Mentat store.
+///
+/// Owned data for the volatile parts (generation and partition map), and `Arc` for the infrequently
+/// changing parts (schema) that we want to share across threads.
+///
+/// See https://github.com/mozilla/mentat/wiki/Thoughts:-modeling-db-conn-in-Rust.
+pub struct Metadata {
+    pub generation: usize,
+    pub partition_map: PartitionMap,
```

Comment:
I'm thinking about this. Is there a situation in which a caller will take a reference to the `partition_map` or `generation`, perhaps in order to validate work optimistically depending on the schema? Does this modeling help or hinder?